src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterDocumentRewriter.cs

```diff
@@ -0,0 +1,695 @@
+                    return varDeclName;
+                }
+
+                var semanticModel = await _originalDocument.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                var semanticFacts = _originalDocument.GetRequiredLanguageService<ISemanticFactsService>();
+                return semanticFacts.GenerateNameForExpression(semanticModel, _expression, capitalize: false, cancellationToken);
+            }
+
+            /// <summary>
+            /// Determines if the expression's grandparent is a variable declarator and if so,
+            /// returns the name
+            /// </summary>
+            private bool ShouldRemoveVariableDeclaratorContainingExpression([NotNullWhen(true)] out string? varDeclName, [NotNullWhen(true)] out SyntaxNode? localDeclaration)
+            {
+                var syntaxFacts = _originalDocument.GetRequiredLanguageService<ISyntaxFactsService>();
+                var declarator = _expression?.Parent?.Parent;
+
+                localDeclaration = null;
+
+                if (!syntaxFacts.IsVariableDeclarator(declarator))
+                {
+                    varDeclName = null;
+                    return false;
+                }
+
+                localDeclaration = Service.GetLocalDeclarationFromDeclarator(declarator);
+                if (localDeclaration is null)
+                {
+                    varDeclName = null;
+                    return false;
+                }
+
+                // TODO: handle in the future
+                if (syntaxFacts.GetVariablesOfLocalDeclarationStatement(localDeclaration).Count > 1)
+                {
+                    varDeclName = null;
+                    localDeclaration = null;
+                    return false;
+                }
+
+                varDeclName = syntaxFacts.GetIdentifierOfVariableDeclarator(declarator).ValueText;
+                return true;
+            }
+
+            public async Task<SyntaxNode> RewriteDocumentAsync(Compilation compilation, Document document, List<SyntaxNode> invocations, CancellationToken cancellationToken)
+            {
+                var syntaxFacts = _originalDocument.GetRequiredLanguageService<ISyntaxFactsService>();
+                var insertionIndex = GetInsertionIndex(compilation, syntaxFacts);
+
+                if (_actionKind is IntroduceParameterCodeActionKind.Overload or IntroduceParameterCodeActionKind.Trampoline)
+                {
+                    return await ModifyDocumentInvocationsTrampolineOverloadAndIntroduceParameterAsync(
+                            compilation, document, invocations, insertionIndex, cancellationToken).ConfigureAwait(false);
+                }
+                else
+                {
+                    return await ModifyDocumentInvocationsAndIntroduceParameterAsync(
+                            compilation, document, insertionIndex, invocations, cancellationToken).ConfigureAwait(false);
+                }
+            }
```

Comment:
consider moving the one public method to the top, below the constructor.  it indicates the only public surface area you are exposing to the caller.

Reply:
Will do